# Problem (2) Data Wine - Classification

### 1. Examine problem 

Menentukan kualitas wine, apakah low atau high, yang akan berpengaruh pada harga jualnya.

### 2. Prepare data 

* Load dataset winequality-white.csv

In [1]:
import pandas as pd

df_wine = pd.read_csv('winequality-white.csv', sep=';')
df_wine.tail()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
4893            6.2              0.21         0.29             1.6      0.039   
4894            6.6              0.32         0.36             8.0      0.047   
4895            6.5              0.24         0.19             1.2      0.041   
4896            5.5              0.29         0.30             1.1      0.022   
4897            6.0              0.21         0.38             0.8      0.020   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
4893                 24.0                  92.0  0.99114  3.27       0.50   
4894                 57.0                 168.0  0.99490  3.15       0.46   
4895                 30.0                 111.0  0.99254  2.99       0.46   
4896                 20.0                 110.0  0.98869  3.34       0.38   
4897                 22.0                  98.0  0.98941  3.26       0.32   

      alcohol  quality  
4893     11.2        6  
4894      9.6        5  
4895      9.4        6  
4896     12.8        7  
4897     11.8        6

* Dataset winequality terdiri dari 12 kolom dan 4898 baris yang semuanya merupakan numeric

In [2]:
df_wine.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4898 entries, 0 to 4897
Data columns (total 12 columns):
fixed acidity           4898 non-null float64
volatile acidity        4898 non-null float64
citric acid             4898 non-null float64
residual sugar          4898 non-null float64
chlorides               4898 non-null float64
free sulfur dioxide     4898 non-null float64
total sulfur dioxide    4898 non-null float64
density                 4898 non-null float64
pH                      4898 non-null float64
sulphates               4898 non-null float64
alcohol                 4898 non-null float64
quality                 4898 non-null int64
dtypes: float64(11), int64(1)
memory usage: 459.3 KB


* Nilai dari kolom label yaitu berupa int64 berkisar dari 3 hingga 9.

In [4]:
df_wine['quality'].value_counts()

6    2198
5    1457
7     880
8     175
4     163
3      20
9       5
Name: quality, dtype: int64

In [5]:
df_wine.shape

(4898, 12)

In [6]:
df_wine['quality'].describe()

count    4898.000000
mean        5.877909
std         0.885639
min         3.000000
25%         5.000000
50%         6.000000
75%         6.000000
max         9.000000
Name: quality, dtype: float64

### Identify Missing Value
* Tidak ada missing value dalam dataset winequality

In [7]:
df_wine.dtypes

fixed acidity           float64
volatile acidity        float64
citric acid             float64
residual sugar          float64
chlorides               float64
free sulfur dioxide     float64
total sulfur dioxide    float64
density                 float64
pH                      float64
sulphates               float64
alcohol                 float64
quality                   int64
dtype: object

In [8]:
df_wine.info

<bound method DataFrame.info of       fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0               7.0             0.270         0.36           20.70      0.045   
1               6.3             0.300         0.34            1.60      0.049   
2               8.1             0.280         0.40            6.90      0.050   
3               7.2             0.230         0.32            8.50      0.058   
4               7.2             0.230         0.32            8.50      0.058   
5               8.1             0.280         0.40            6.90      0.050   
6               6.2             0.320         0.16            7.00      0.045   
7               7.0             0.270         0.36           20.70      0.045   
8               6.3             0.300         0.34            1.60      0.049   
9               8.1             0.220         0.43            1.50      0.044   
10              8.1             0.270         0.41            1.45      0.033

### Identify Outlier
* High outlier ditemukan di hampir semua kolom kecuali kolom 'alcohol'
* Low outlier ditemukan di beberapa kolom, yaitu 'fixed acidity', 'citrid acid', 'chlorides', 'total sulfur dioxide', 'pH', dan 'quality'
* Pada kolom 'alcohol' tidak ditemukan adanya outlier, baik itu low outlier maupun high outlier

In [9]:
for i in range(len(df_wine.columns)):
    if (df_wine[df_wine.columns[i]].dtypes in ['int64','float64']):
        print('\nAttribute-',df_wine.columns[i],':',df_wine[df_wine.columns[i]].dtypes)
        Q1=df_wine[df_wine.columns[i]].quantile(0.25)
        print('Q1',Q1)
        Q3=df_wine[df_wine.columns[i]].quantile(0.75)
        print('Q3',Q3)
        IQR=Q3-Q1
        print('IQR',IQR)
        min=df_wine[df_wine.columns[i]].min()
        max=df_wine[df_wine.columns[i]].max()
        min_IQR=Q1-1.5*IQR
        max_IQR=Q3+1.5*IQR
        if (min<min_IQR):
            print('Low outlier is found < ',min_IQR,' - ', min)
        if (max>max_IQR):
            print('High outlier is found > ',max_IQR,' - ', max)


Attribute- fixed acidity : float64
Q1 6.3
Q3 7.3
IQR 1.0
Low outlier is found <  4.8  -  3.8
High outlier is found >  8.8  -  14.2

Attribute- volatile acidity : float64
Q1 0.21
Q3 0.32
IQR 0.11000000000000001
High outlier is found >  0.48500000000000004  -  1.1

Attribute- citric acid : float64
Q1 0.27
Q3 0.39
IQR 0.12
Low outlier is found <  0.09000000000000002  -  0.0
High outlier is found >  0.5700000000000001  -  1.66

Attribute- residual sugar : float64
Q1 1.7
Q3 9.9
IQR 8.200000000000001
High outlier is found >  22.200000000000003  -  65.8

Attribute- chlorides : float64
Q1 0.036000000000000004
Q3 0.05
IQR 0.013999999999999999
Low outlier is found <  0.015000000000000006  -  0.009000000000000001
High outlier is found >  0.07100000000000001  -  0.34600000000000003

Attribute- free sulfur dioxide : float64
Q1 23.0
Q3 46.0
IQR 23.0
High outlier is found >  80.5  -  289.0

Attribute- total sulfur dioxide : float64
Q1 108.0
Q3 167.0
IQR 59.0
Low outlier is found <  19.5  -  9.0
High

### Drop outliers which have big gap with others
* Untuk kolom yang memiliki high outlier yang jaraknya sangat jauh dari tetangga terdekatnya akan dilakukan proses drop agar sebarannya tidak terlalu jauh
* Kolom-kolom tersebut yaitu 'free sulfur dioxide', 'residual sugar', dan 'total sulfur dioxide'
* Untuk masing-masing kolom tersebut hanya ada satu instance yang di drop
* Outlier terjauh untuk kolom 'free sulfur dioxide' dan 'total sulfur dioxide' berada di instance yang sama yaitu di baris 4745, sedangkan outlier terjauh untuk kolom 'residual sugar' berada di baris 2781, sehingga untuk data winequaality ini akan dihapus sebanyak dua instance

In [10]:
df_wine.describe()

fixed acidity  volatile acidity  citric acid  residual sugar  \
count    4898.000000       4898.000000  4898.000000     4898.000000   
mean        6.854788          0.278241     0.334192        6.391415   
std         0.843868          0.100795     0.121020        5.072058   
min         3.800000          0.080000     0.000000        0.600000   
25%         6.300000          0.210000     0.270000        1.700000   
50%         6.800000          0.260000     0.320000        5.200000   
75%         7.300000          0.320000     0.390000        9.900000   
max        14.200000          1.100000     1.660000       65.800000   

         chlorides  free sulfur dioxide  total sulfur dioxide      density  \
count  4898.000000          4898.000000           4898.000000  4898.000000   
mean      0.045772            35.308085            138.360657     0.994027   
std       0.021848            17.007137             42.498065     0.002991   
min       0.009000             2.000000              9.000000     0.987110   
25%       0.036000            23.000000            108.000000     0.991723   
50%       0.043000            34.000000            134.000000     0.993740   
75%       0.050000            46.000000            167.000000     0.996100   
max       0.346000           289.000000            440.000000     1.038980   

                pH    sulphates      alcohol      quality  
count  4898.000000  4898.000000  4898.000000  4898.000000  
mean      3.188267     0.489847    10.514267     5.877909  
std       0.151001     0.114126     1.230621     0.885639  
min       2.720000     0.220000     8.000000     3.000000  
25%       3.090000     0.410000     9.500000     5.000000  
50%       3.180000     0.470000    10.400000     6.000000  
75%       3.280000     0.550000    11.400000     6.000000  
max       3.820000     1.080000    14.200000     9.000000

In [11]:
df_wine[df_wine['free sulfur dioxide']>=200]

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
4745            6.1              0.26         0.25             2.9      0.047   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
4745                289.0                 440.0  0.99314  3.44       0.64   

      alcohol  quality  
4745     10.5        3

In [12]:
df_wine[df_wine['residual sugar']>=35]

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
2781            7.8             0.965          0.6            65.8      0.074   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
2781                  8.0                 160.0  1.03898  3.39       0.69   

      alcohol  quality  
2781     11.7        6

In [13]:
df_wine[df_wine['total sulfur dioxide']>=400]

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
4745            6.1              0.26         0.25             2.9      0.047   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
4745                289.0                 440.0  0.99314  3.44       0.64   

      alcohol  quality  
4745     10.5        3

In [14]:
df_wine = df_wine.drop([2781, 4745])

In [15]:
df_wine

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0               7.0             0.270         0.36           20.70      0.045   
1               6.3             0.300         0.34            1.60      0.049   
2               8.1             0.280         0.40            6.90      0.050   
3               7.2             0.230         0.32            8.50      0.058   
4               7.2             0.230         0.32            8.50      0.058   
5               8.1             0.280         0.40            6.90      0.050   
6               6.2             0.320         0.16            7.00      0.045   
7               7.0             0.270         0.36           20.70      0.045   
8               6.3             0.300         0.34            1.60      0.049   
9               8.1             0.220         0.43            1.50      0.044   
10              8.1             0.270         0.41            1.45      0.033   
11              8.6             0.230         0.40            4.20      0.035   
12              7.9             0.180         0.37            1.20      0.040   
13              6.6             0.160         0.40            1.50      0.044   
14              8.3             0.420         0.62           19.25      0.040   
15              6.6             0.170         0.38            1.50      0.032   
16              6.3             0.480         0.04            1.10      0.046   
17              6.2             0.660         0.48            1.20      0.029   
18              7.4             0.340         0.42            1.10      0.033   
19              6.5             0.310         0.14            7.50      0.044   
20              6.2             0.660         0.48            1.20      0.029   
21              6.4             0.310         0.38            2.90      0.038   
22              6.8             0.260         0.42            1.70      0.049   
23              7.6             0.670         0.14            1.50      0.074   
24              6.6             0.270         0.41            1.30      0.052   
25              7.0             0.250         0.32            9.00      0.046   
26              6.9             0.240         0.35            1.00      0.052   
27              7.0             0.280         0.39            8.70      0.051   
28              7.4             0.270         0.48            1.10      0.047   
29              7.2             0.320         0.36            2.00      0.033   
...             ...               ...          ...             ...        ...   
4868            5.8             0.230         0.31            4.50      0.046   
4869            6.6             0.240         0.33           10.10      0.032   
4870            6.1             0.320         0.28            6.60      0.021   
4871            5.0             0.200         0.40            1.90      0.015   
4872            6.0             0.420         0.41           12.40      0.032   
4873            5.7             0.210         0.32            1.60      0.030   
4874            5.6             0.200         0.36            2.50      0.048   
4875            7.4             0.220         0.26            1.20      0.035   
4876            6.2             0.380         0.42            2.50      0.038   
4877            5.9             0.540         0.00            0.80      0.032   
4878            6.2             0.530         0.02            0.90      0.035   
4879            6.6             0.340         0.40            8.10      0.046   
4880            6.6             0.340         0.40            8.10      0.046   
4881            5.0             0.235         0.27           11.75      0.030   
4882            5.5             0.320         0.13            1.30      0.037   
4883            4.9             0.470         0.17            1.90      0.035   
4884            6.5             0.330         0.38            8.30      0.048   
4885            6.6             0.340

### Binning process
* proses binning dilakukan untuk membagi label di kolom 'quality' menjadi dua kategori yaitu low dan high
* low quality yaitu terdiri dari label 1 sampai 5 dan direpresentasikan dengan angka 0, sedangkan high quality terdiri dari label 6 sampai 10 dan direpresentasikan dengan angka 1

In [16]:
bins = [1, 6, 10]
quality_cat=[0, 1]
df_wine['quality_category'] = pd.cut(df_wine['quality'], bins, right=False, labels=quality_cat)
df_wine.sample(10)

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
1369            7.4              0.64         0.47           14.15      0.168   
4576            5.9              0.30         0.29            1.10      0.036   
4087            6.5              0.21         0.42            1.10      0.059   
262             6.9              0.28         0.40            8.20      0.036   
1734            6.9              0.16         0.30            9.60      0.057   
1200            6.8              0.18         0.37            1.50      0.027   
3070            6.8              0.28         0.43            7.60      0.030   
3329            6.7              0.21         0.34            1.40      0.049   
1488            6.5              0.26         0.74           13.30      0.044   
3492            6.0              0.28         0.27           15.50      0.036   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
1369                 42.0                 185.0  0.99840  2.90       0.49   
4576                 23.0                  56.0  0.99040  3.19       0.38   
4087                 33.0                 101.0  0.99270  3.12       0.38   
262                  15.0                  95.0  0.99440  3.17       0.33   
1734                 50.0                 185.0  0.99780  3.39       0.38   
1200                 37.0                  93.0  0.99200  3.30       0.45   
3070                 30.0                 110.0  0.99164  3.08       0.59   
3329                 36.0                 112.0  0.99091  3.02       0.50   
1488                 68.0                 224.0  0.99720  3.18       0.54   
3492                 31.0                 134.0  0.99408  3.19       0.44   

      alcohol  quality quality_category  
1369      9.3        5                0  
4576     11.3        5                0  
4087      9.7        6                1  
262      10.2        5                0  
1734      9.6        6                1  
1200     10.8        6                1  
3070     12.5        8                1  
3329     11.0        6                1  
1488      9.5        6                1  
3492     13.0        7                1

* Jumlah instance yang termasuk low quality sebanyak 1639, dan yang termasuk high quality sebanyak 3257

In [17]:
df_wine['quality_category'].value_counts()

1    3257
0    1639
Name: quality_category, dtype: int64

In [18]:
df_wine.dtypes

fixed acidity            float64
volatile acidity         float64
citric acid              float64
residual sugar           float64
chlorides                float64
free sulfur dioxide      float64
total sulfur dioxide     float64
density                  float64
pH                       float64
sulphates                float64
alcohol                  float64
quality                    int64
quality_category        category
dtype: object

### 3. Prepare the experiments
#### Tujuan Penelitian
* Membandingkan model KNN classifier dengan parameter k=50, k=20, k=10, dan k=5 dengan untuk setiap parameter diterapkan metode SMOTE (Synthetic Minority Over-sampling Technique) dan tidak diterapkan metode SMOTE.
* Membandingkan Membandingkan model Naive Bayes classifier dengan menerapkan metode SMOTE untuk menangani imbalanced data dan tidak diterapkan metode SMOTE.
* Membandingkan model classifier yang lebih baik diantara dua classifier yang diujikan untuk data wine.
#### Respon / Ukuran Kinerja
* Persentase wine diklasifikasikan dengan benar, apakah termasuk low quality atau high quality.
#### Faktor dan Level
* Faktor yang mempengaruhi untuk penerapan algoritma KNN adalah nilai k, atau jumlah tetangga terdekat yang akan diambil.
* Level nilai k ditentukan tidak berurutan.
#### Desain Eksperimen
* Strategi eksperimen untuk dataset wine menggunakan strategi Best Guess untuk nilai k.
* Skema eksperimen yang digunakan adalah Hold-out validation yaitu membagi datset menjadi data training dan data testing.

### 4. Modeling

### Membagi dataset menjadi data training dan data testing

In [19]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

source = df_wine.iloc[:,:-1] #all columns except the last one
target = df_wine['quality_category'] #only the last column

sv_train, sv_test, tv_train, tv_test = train_test_split(source, target, random_state=3, test_size=0.2)
print("\nData Training:\n")
print(sv_train.head())
print(sv_train.shape)

print("\nData Testing:\n")
print(sv_test.head())
print(sv_test.shape)


Data Training:

      fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
445             7.1              0.32         0.32            11.0      0.038   
563             7.4              0.30         0.22             1.4      0.046   
3000            6.7              0.18         0.28            10.2      0.039   
4148            6.2              0.20         0.25            15.0      0.055   
4178            5.8              0.19         0.33             4.2      0.038   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
445                  16.0                  66.0  0.99370  3.24       0.40   
563                  16.0                 135.0  0.99280  3.08       0.77   
3000                 29.0                 115.0  0.99469  3.11       0.45   
4148                  8.0                 120.0  0.99767  3.19       0.53   
4178                 49.0                 133.0  0.99107  3.16       0.42   

      alcohol  quality  
445     

### Evaluasi score tanpa meng-handle imbalanced data

#### KNN Classifier (k=50)

In [20]:
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score
from sklearn.metrics import f1_score

knn_model = KNeighborsClassifier(n_neighbors=50)

knn_model.fit(sv_train, tv_train)
pred = knn_model.predict(sv_test)
print("[INFO] evaluating...")
print(classification_report(tv_test, pred))

[INFO] evaluating...
              precision    recall  f1-score   support

           0       0.57      0.33      0.42       327
           1       0.72      0.88      0.79       653

    accuracy                           0.69       980
   macro avg       0.65      0.60      0.60       980
weighted avg       0.67      0.69      0.67       980



#### KNN Classifier (k=20)

In [21]:
knn_model = KNeighborsClassifier(n_neighbors=20)

knn_model.fit(sv_train, tv_train)
pred = knn_model.predict(sv_test)
print("[INFO] evaluating...")
print(classification_report(tv_test, pred))

[INFO] evaluating...
              precision    recall  f1-score   support

           0       0.58      0.43      0.50       327
           1       0.75      0.84      0.79       653

    accuracy                           0.71       980
   macro avg       0.67      0.64      0.65       980
weighted avg       0.69      0.71      0.69       980



#### KNN Classifier (k=10)

In [22]:
knn_model = KNeighborsClassifier(n_neighbors=10)

knn_model.fit(sv_train, tv_train)
pred = knn_model.predict(sv_test)
print("[INFO] evaluating...")
print(classification_report(tv_test, pred))

[INFO] evaluating...
              precision    recall  f1-score   support

           0       0.62      0.53      0.57       327
           1       0.78      0.84      0.81       653

    accuracy                           0.73       980
   macro avg       0.70      0.68      0.69       980
weighted avg       0.73      0.73      0.73       980



#### KNN Classifier (k=5)

In [23]:
knn_model = KNeighborsClassifier(n_neighbors=5)

knn_model.fit(sv_train, tv_train)
pred = knn_model.predict(sv_test)
print("[INFO] evaluating...")
print(classification_report(tv_test, pred))

[INFO] evaluating...
              precision    recall  f1-score   support

           0       0.65      0.52      0.57       327
           1       0.78      0.86      0.82       653

    accuracy                           0.74       980
   macro avg       0.71      0.69      0.70       980
weighted avg       0.73      0.74      0.74       980



#### KNN Classifier (k=3)

In [24]:
knn_model = KNeighborsClassifier(n_neighbors=3)

knn_model.fit(sv_train, tv_train)
pred = knn_model.predict(sv_test)
print("[INFO] evaluating...")
print(classification_report(tv_test, pred))

[INFO] evaluating...
              precision    recall  f1-score   support

           0       0.64      0.56      0.60       327
           1       0.79      0.85      0.82       653

    accuracy                           0.75       980
   macro avg       0.72      0.70      0.71       980
weighted avg       0.74      0.75      0.75       980



#### Naive Bayes

In [25]:
from sklearn.naive_bayes import GaussianNB

nb_model = GaussianNB()

nb_model.fit(sv_train, tv_train)
pred = nb_model.predict(sv_test)
print("[INFO] evaluating...")
print(classification_report(tv_test, pred))

[INFO] evaluating...
              precision    recall  f1-score   support

           0       0.97      0.94      0.96       327
           1       0.97      0.98      0.98       653

    accuracy                           0.97       980
   macro avg       0.97      0.96      0.97       980
weighted avg       0.97      0.97      0.97       980



### Handling imbalanced data with oversampling method - SMOTE

In [26]:
from imblearn.over_sampling import SMOTE

tv_train.value_counts()

Using TensorFlow backend.


1    2604
0    1312
Name: quality_category, dtype: int64

#### Proses Oversampling SMOTE

In [27]:
smt = SMOTE()
sv_train, tv_train = smt.fit_sample(sv_train, tv_train)

#### Duplikasi nilai label minoritas menjadi sama dengan label mayoritas, sehingga untuk setiap label nilainya sekarang menjadi 2604

In [28]:
np.bincount(tv_train)

array([2604, 2604], dtype=int64)

### Evaluasi score dengan menerapkan Oversampling SMOTE

#### KNN Classifier (k=50) SMOTE

In [29]:
knn_model = KNeighborsClassifier(n_neighbors=50)

knn_model.fit(sv_train, tv_train)
pred = knn_model.predict(sv_test)
print("[INFO] evaluating...")
print(classification_report(tv_test, pred))

[INFO] evaluating...
              precision    recall  f1-score   support

           0       0.49      0.72      0.58       327
           1       0.82      0.62      0.70       653

    accuracy                           0.65       980
   macro avg       0.65      0.67      0.64       980
weighted avg       0.71      0.65      0.66       980



#### KNN Classifier (k=20) SMOTE

In [30]:
knn_model = KNeighborsClassifier(n_neighbors=20)

knn_model.fit(sv_train, tv_train)
pred = knn_model.predict(sv_test)
print("[INFO] evaluating...")
print(classification_report(tv_test, pred))

[INFO] evaluating...
              precision    recall  f1-score   support

           0       0.52      0.74      0.61       327
           1       0.84      0.65      0.73       653

    accuracy                           0.68       980
   macro avg       0.68      0.70      0.67       980
weighted avg       0.73      0.68      0.69       980



#### KNN Classifier (k=10) SMOTE

In [31]:
knn_model = KNeighborsClassifier(n_neighbors=10)

knn_model.fit(sv_train, tv_train)
pred = knn_model.predict(sv_test)
print("[INFO] evaluating...")
print(classification_report(tv_test, pred))

[INFO] evaluating...
              precision    recall  f1-score   support

           0       0.52      0.75      0.62       327
           1       0.84      0.66      0.74       653

    accuracy                           0.69       980
   macro avg       0.68      0.70      0.68       980
weighted avg       0.73      0.69      0.70       980



#### KNN Classifier (k=5) SMOTE

In [32]:
knn_model = KNeighborsClassifier(n_neighbors=5)

knn_model.fit(sv_train, tv_train)
pred = knn_model.predict(sv_test)
print("[INFO] evaluating...")
print(classification_report(tv_test, pred))

[INFO] evaluating...
              precision    recall  f1-score   support

           0       0.58      0.72      0.64       327
           1       0.84      0.74      0.79       653

    accuracy                           0.73       980
   macro avg       0.71      0.73      0.71       980
weighted avg       0.75      0.73      0.74       980



#### KNN Classifier (k=3) SMOTE

In [33]:
knn_model = KNeighborsClassifier(n_neighbors=3)

knn_model.fit(sv_train, tv_train)
pred = knn_model.predict(sv_test)
print("[INFO] evaluating...")
print(classification_report(tv_test, pred))

[INFO] evaluating...
              precision    recall  f1-score   support

           0       0.60      0.69      0.64       327
           1       0.83      0.77      0.80       653

    accuracy                           0.74       980
   macro avg       0.71      0.73      0.72       980
weighted avg       0.75      0.74      0.74       980



#### Naive Bayes - SMOTE

In [34]:
from sklearn.naive_bayes import GaussianNB

nb_model = GaussianNB()

nb_model.fit(sv_train, tv_train)
pred = nb_model.predict(sv_test)
print("[INFO] evaluating...")
print(classification_report(tv_test, pred))

[INFO] evaluating...
              precision    recall  f1-score   support

           0       0.98      0.95      0.97       327
           1       0.98      0.99      0.98       653

    accuracy                           0.98       980
   macro avg       0.98      0.97      0.97       980
weighted avg       0.98      0.98      0.98       980



### Kesimpulan
* Untuk Algoritma KNN tanpa diterapkan metode SMOTE untuk menangani imbalanced data, nilai parameter k yang paling baik adalah saat k=3.
* Untuk Algoritma KNN dengan menerapkan metode SMOTE untuk menangani imbalanced data, nilai parameter k yang paling baik adalah saat k=5.
* Algoritma KNN untuk setiap parameter k, score terbaik adalah dengan menerapkan metode SMOTE untuk menangani imbalanced data.
* Begitu pula untuk algoritma Naive Bayes, score terbaik adalah ketika menerapkan metode SMOTE untuk menangani imbalanced data.
* Dari dua algoritma yang digunakan dalam eksperimen untuk data wine quality, score yang lebih baik adalah ketika menggunakan algoritma Naive Bayes dan menerapkan metode oversampling SMOTE.